pkg/controller/infrastructure/infraflow/reconciler.go

```diff
@@ -0,0 +1,158 @@
+	"github.com/gardener/gardener-extension-provider-gcp/pkg/internal/infrastructure"
+)
+
+// FlowReconciler is capable of reconciling and deleting the infrastructure for a shoot.
+type FlowReconciler struct {
+	*shared.BasicFlowContext
+
+	infra          *extensionsv1alpha1.Infrastructure
+	config         *gcp.InfrastructureConfig
+	updater        gcpclient.Updater
+	serviceAccount *gcpinternal.ServiceAccount
+	clusterName    string
+	whiteboard     shared.Whiteboard
+	podCIDR        *string
+
+	computeClient gcpclient.ComputeClient
+	iamClient     gcpclient.IAMClient
+}
+
+// NewFlowReconciler returns a new FlowReconciler.
+func NewFlowReconciler(
+	ctx context.Context,
+	log logr.Logger,
+	infra *extensionsv1alpha1.Infrastructure,
+	cluster *controller.Cluster,
+	c client.Client,
+) (*FlowReconciler, error) {
+	config, err := helper.InfrastructureConfigFromInfrastructure(infra)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceAccount, err := gcpinternal.GetServiceAccountFromSecretReference(ctx, c, infra.Spec.SecretRef)
+	if err != nil {
+		return nil, err
+	}
+
+	gc := gcpclient.New()
+	com, err := gc.Compute(ctx, c, infra.Spec.SecretRef)
+	if err != nil {
+		return nil, err
+	}
+
+	iam, err := gc.IAM(ctx, c, infra.Spec.SecretRef)
+	if err != nil {
+		return nil, err
+	}
+
+	wb := shared.NewWhiteboard()
+	bfc := shared.NewBasicFlowContext(log, wb, nil)
+	fr := &FlowReconciler{
+		BasicFlowContext: bfc,
+		whiteboard:       wb,
+		infra:            infra,
+		serviceAccount:   serviceAccount,
+		config:           config,
+		updater:          gcpclient.NewUpdater(gc, serviceAccount, log),
+		clusterName:      cluster.ObjectMeta.Name,
+		podCIDR:          cluster.Shoot.Spec.Networking.Pods,
+
+		computeClient: com,
+		iamClient:     iam,
+	}
+
+	return fr, nil
+}
+
+// Reconcile reconciles the infrastructure
+func (c *FlowReconciler) Reconcile(ctx context.Context) (*v1alpha1.InfrastructureStatus, *runtime.RawExtension, error) {
+	g := c.buildReconcileGraph()
+	f := g.Compile()
+	c.Log.Info("starting Flow Reconciliation")
+	err := f.Run(ctx, flow.Opts{Log: c.Log})
+	if err != nil {
+		c.Log.Error(err, "flow reconciliation failed")
+		status, state, inErr := c.getStatus()
+		if inErr != nil {
+			c.Log.Error(err, "flow reconciliation failed")
+		}
+		return status, state, err
+	}
+
+	return c.getStatus()
+}
+
+// Delete is used to destroy the infrastructure.
+func (c *FlowReconciler) Delete(ctx context.Context) error {
+	g := c.buildDeleteGraph()
+	f := g.Compile()
+	return f.Run(ctx, flow.Opts{Log: c.Log})
+}
+
+func (c *FlowReconciler) getStatus() (*v1alpha1.InfrastructureStatus, *runtime.RawExtension, error) {
+	status := &v1alpha1.InfrastructureStatus{
+		TypeMeta: infrastructure.StatusTypeMeta,
+		Networks: v1alpha1.NetworkStatus{
+			VPC:     v1alpha1.VPC{},
+			Subnets: []v1alpha1.Subnet{},
+			NatIPs:  []v1alpha1.NatIP{},
+		},
+	}
+
+	if n := GetObject[*gcpclient.Network](c.whiteboard, ObjectKeyVPC); n != nil {
```

Comment:
If the object is not on the map, `GetObject` will always return an instantiated obj of the templated type. In this case because I store and and retrieve pointers it will be a "zero value" pointer which can be nil.